### ISSUE TITLE
Feature request: get a list of premoves

### ISSUE THREAD

**Issue by alexobviously** [jhlywa/chess.js#287]
It would be nice to have a function that gets the available premoves for a position - moves that the player who isn't currently playing could potentially play on their next move, regardless of the move that the active player plays.

I don't really do JS but I have implemented the same thing in chess960.dart (based on chess.dart which is a port of this package). See here: https://github.com/alexobviously/chess960.dart/blob/master/lib/chess960.dart#L1313

Some modifications to move generation would be required, see the parts about `allMoves` in [generate_moves](https://github.com/alexobviously/chess960.dart/blob/master/lib/chess960.dart#L611)

Currently the only way to do this with this package, afaik, is to loop through all of the squares and get all the moves manually, which is kind of a pain, especially when you want the premoves separate. So it is possible to do already, but it would be a nice feature to have if there was a distinct function for it.

**Comment by yanush** [comment]
I would really like this feature too. I think it's super important since many chess apps show you indicators for legal moves, but legal pre-moves are different. Any easy way to achieve that without going though all possible moves for every type of piece?

**Comment by neofight78** [comment]
@alexobviously @yanush  If we are talking about pre-moves, do you really want all possible moves, or just moves for one specific piece?

**Comment by yanush** [comment]
I'm no longer working on this project, but I needed that for a specific piece.
In my chosen Lichess theme for example, the indicators for legal moves are green dots on the valid squares, while the indicators for legal pre-moves are purple dots on the valid squares.
Purple dots appear when you click a piece and it's not your turn
Green dots appear when you click a piece when it's your turn

**Comment by neofight78** [comment]
Thanks for the reply! Do you know the algorithm that should be used or what you were looking for? Just experimented a bit in a blitz, it looks like you can choose any square as long as it's not blocked by your own colour, and you can capture that first piece of your own colour (but can travel not further). These seems to be true even if those are not possible for any move that your opponent might make.

**Comment by neofight78** [comment]
chess.com seems even more relaxed, it ignores what pieces are on the board. I presume this is because they allow multiple pre-moves and so they can't really restrict the moves available as the number of possibilities explodes.

**Comment by yanush** [comment]
I don't think the algorithm should give you 100% invalid moves as a pre-move option
for example, it shouldn't let you put a white bishop on a dark square
it should let you jump with your nights to an invalid square
if i had to come up with an algorithm for valid pre-moves, i would probably assume that the piece i want to move is the only piece on the board, and then all its legal moves are valid as pre-moves
